kclvm/tools/src/LSP/src/main.rs

```diff
@@ -0,0 +1,191 @@
+            },
+        })
+    }
+
+    async fn initialized(&self, _: InitializedParams) {
+        self.client
+            .log_message(MessageType::INFO, "initialized!")
+            .await;
+    }
+
+    async fn shutdown(&self) -> Result<()> {
+        Ok(())
+    }
+
+    async fn did_open(&self, params: DidOpenTextDocumentParams) {
+        self.client
+            .log_message(MessageType::INFO, "file opened!")
+            .await;
+        self.on_change(TextDocumentItem {
+            uri: params.text_document.uri,
+        })
+        .await
+    }
+
+    async fn did_change(&self, params: DidChangeTextDocumentParams) {
+        self.on_change(TextDocumentItem {
+            uri: params.text_document.uri,
+        })
+        .await;
+        self.client
+            .log_message(MessageType::INFO, "file changed!")
+            .await;
+    }
+
+    async fn did_save(&self, params: DidSaveTextDocumentParams) {
+        self.on_change(TextDocumentItem {
+            uri: params.text_document.uri,
+        })
+        .await;
+
+        self.client
+            .log_message(MessageType::INFO, "file saved!")
+            .await;
+    }
+
+    async fn did_close(&self, _: DidCloseTextDocumentParams) {
+        self.client
+            .log_message(MessageType::INFO, "file closed!")
+            .await;
+    }
+}
+
+impl Backend {
+    async fn on_change(&self, params: TextDocumentItem) {
+        self.client
+            .log_message(
+                MessageType::INFO,
+                format!(
+                    "Get request: {} ",
+                    Local
+                        .timestamp_millis_opt(Local::now().timestamp_millis())
+                        .unwrap()
+                ),
+            )
+            .await;
+        let uri = params.uri.clone();
+        let file_name = uri.path();
+        self.client
+            .log_message(MessageType::INFO, "on change")
+            .await;
+
+        self.client
+            .log_message(
+                MessageType::INFO,
+                format!(
+                    "Start lint: {} ",
+                    Local
+                        .timestamp_millis_opt(Local::now().timestamp_millis())
+                        .unwrap()
+                ),
+            )
+            .await;
+
+        let (errors, warnings) = lint_files(&[file_name], None);
+
+        self.client
+            .log_message(
+                MessageType::INFO,
+                format!(
+                    "End lint: {} ",
+                    Local
+                        .timestamp_millis_opt(Local::now().timestamp_millis())
+                        .unwrap()
+                ),
+            )
+            .await;
+        let diags: IndexSet<KCLDiagnostic> = errors
+            .iter()
+            .chain(warnings.iter())
+            .cloned()
+            .collect::<IndexSet<KCLDiagnostic>>();
+
+        let mut diagnostics: Vec<Diagnostic> = vec![];
```

Comment:
It is necessary to implement the Into trait from KCLDiagnostic to LSPDiagnostic to achieve error reuse of different events.